### DISCUSSION TITLE
cors origin problem

### DISCUSSION THREAD

**Discussion by TrabelsiHoussem** [emqx/emqx#12073] [Q&A]
even i set the dashboard.cors = true, i always got an error of cors origin below is the error :
has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials' mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.

**Comment by SergeTupchiy** [comment]
Hi @TrabelsiHoussem, 

When dashboard.cors = true, EMQX adds `Access-Control-Allow-Origin: *` response header for any origin.
You can probably set withCredentials to false as explained here: https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS/Errors/CORSNotSupportingCredentials

**Reply by TrabelsiHoussem** [reply]
when i set the withCredentials to true  i get :
Access to XMLHttpRequest at 'http://localhost:18083/api/v5/clients' from origin 'http://localhost:8100' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: The value of the 'Access-Control-Allow-Origin' header in the response must not be the wildcard '*' when the request's credentials mode is 'include'. The credentials mode of requests initiated by the XMLHttpRequest is controlled by the withCredentials attribute.

when withCredentials is false i get : 

Access to XMLHttpRequest at 'http://localhost:18083/api/v5/clients' from origin 'http://localhost:8100' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.

**Comment by ysfscream** [comment]
@TrabelsiHoussem Hello,

First, I'd like to confirm your development scenario. Are you running a service at http://localhost:8100/ and then accessing the EMQX API? Have you considered adding a reverse proxy to your service? Since EMQX is based on JWT authentication, you might try turning off `withCredentials`. However, both methods are causing errors. I would suggest using an API client to access EMQX and checking if your service doesn't have the necessary settings. Thank you.

**Comment by dwcahyo23** [comment]
i have same issue with emqx 5.8.1

Access to XMLHttpRequest at 'http://192.168.192.7:18083/api/v5/publish' from origin 'http://localhost:5173' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource.

**Comment by id** [comment]
EMQX correctly sets the `Access-Control-Allow-Origin: *` header in response when

1. The `Origin` header is present in the request
2. `dashboard.cors` is set to `true`

It's important that both of these conditions are met.

```
curl -v -H "Authorization: Bearer $EMQX_TOKEN" \
            -H "Content-type: application/json" \
            -H "Origin: http://localhost" \
            "http://127.0.0.1:18083/api/v5/publish" \
            -d '{"topic": "foo", "payload": "bar"}'
*   Trying 127.0.0.1:18083...
* Connected to 127.0.0.1 (127.0.0.1) port 18083
> POST /api/v5/publish HTTP/1.1
> Host: 127.0.0.1:18083
> User-Agent: curl/8.7.1
> Accept: */*
> Authorization: Bearer ***
> Content-type: application/json
> Origin: http://localhost
> Content-Length: 34
>
* upload completely sent off: 34 bytes
< HTTP/1.1 202 Accepted
< Access-Control-Allow-Origin: *
< content-length: 54
< content-type: application/json
< date: Wed, 06 Nov 2024 07:20:35 GMT
< server: Cowboy
<
* Connection #0 to host 127.0.0.1 left intact
{"message":"no_matching_subscribers","reason_code":16}
```
